scripts/Valves/valve.py

```diff
@@ -0,0 +1,448 @@
+#!/usr/bin/env python3
+
+import math
+import os
+import sys
+
+# load parent path of KicadModTree
+sys.path.append(os.path.join(sys.path[0], "..", ".."))
+
+from KicadModTree import *
+        
+        
+def create_sadle(f, origo_x, origo_y, sadle, sadle_hole, sadle_pcb_hole, nlayer, wLayer):
+
+    #
+    # sadle = [6.6, 35.5, 26.5, 14.25, 13.2, 72],        # sadle z pos, length, width, xpos r2, diameter r2, rotation
+    #
+    sadle_z = sadle[0]
+    sadle_w = sadle[1]
+    sadle_r1 = sadle[2] / 2.0
+    sadle_x = sadle[3]
+    sadle_r2 = sadle[4] / 2.0
+    sadle_a = sadle[5]
+    sadle_h = 0.2
+    #
+    # Dummy
+#    f.append(Circle(center=(round(origo_x, 2), round(origo_y, 2)), radius=round(sadle_r1, 2), layer=nlayer, width=wLayer))
+    #
+#    f.append(Circle(center=(round(origo_x + sadle_x, 2), round(origo_y, 2)), radius=round(sadle_r2, 2), layer=nlayer, width=wLayer))
+    #
+#    f.append(Circle(center=(round(origo_x - sadle_x, 2), round(origo_y, 2)), radius=round(sadle_r2, 2), layer=nlayer, width=wLayer))
+    #
+    # https://en.wikipedia.org/wiki/Tangent_lines_to_circles
+    #
+    x1 = 0.0 - sadle_x;
+    y1 = 0.0
+    r = sadle_r2
+    x2 = 0.0
+    y2 = 0.0
+    R = sadle_r1
+    #
+    theta = 0 - math.atan((y2 - y1) / (x2 - x1))
+    beta = math.asin((R - r) / (math.sqrt(((x2 - x1)**2) + ((y2 - y1)**2))))
+    alpha = theta - beta
+    x3 = x1 + r * math.cos((math.pi / 2.0) - alpha)
+    y3 = y1 + r * math.sin((math.pi / 2.0) - alpha)
+    x4 = x2 + R * math.cos((math.pi / 2.0) - alpha)
+    y4 = y2 + R * math.sin((math.pi / 2.0) - alpha)
+  
+    #
+    th = math.sqrt((((0.0 - sadle_x) - x3)**2) + ((0.0 - y3)**2))
+    aa = math.fabs(math.degrees(math.acos((x3 - (0.0 - sadle_x)) / th))) - 90.0
+    #
+    # reflect Y in small circle
+    #
+    y5 = 0.0 - (y3 - 0.0)
+    y6 = 0.0 - (y4 - 0.0)
+    #
+    x5 = x3
+    x6 = x4
+    # Upper left line
+    #
+    # Upper part of large circle
+    th = math.sqrt(((0.0 - x6)**2) + ((0.0 - y6)**2))
+    aa2 = 90.0 - math.fabs(math.degrees(math.acos((0.0 - x6) / th)))
+    #
+    # reflect X in large circle
+    #
+    x7 = 0.0 + (0.0 - x6)
+    y7 = y6
+    #
+    # reflect small circle in the large circles origo
+    #
+    x8 = 0.0 + (0.0 - x5)
+    y8 = y5
+    #
+    #
+    # reflect small circle in the large circles origo
+    #
+    #
+    # reflect large circle in the large circles origo
+    #
+    x10 = x7
+    y10 = 0.0 + (0.0 - y7)
+    x9 = x8
+    y9 = y3
```

Comment:
To have an array for below you could do the following:
`coords=[None]*10`
and then instead of writing `x9 = x8`, `y9 = y3` you would write `coords[8] = Vector2D(coords[7].x, coords[2].y)`

(One does not simply start to count with x1,y1 it is x0,y0 :-) )

Reply:
"(One does not simply start to count with x1,y1 it is x0,y0 :-)"

ohhh, you should have been around the time when printers almost could not distingues between 0 and O :)